### DISCUSSION TITLE
New API Portal Authorize Button showing flow: undefined

### DISCUSSION THREAD

**Discussion by phil-hotchkiss** [mia-platform/community#518] [Q&A] [answered]
### What product are you having troubles with?

Console

### What Console version are you using?

1.102.8

### Description

Hi there! I recently upgraded to the latest versions of swagger-aggregator and api portal in order to take advantage of the new `Authorize` button that is available instead of having to get an authorization code outside of the api portal experience like in previous versions. 

I have the following code block that is specified in my swagger-aggregator.json
```json
{
  "baseSwagger": {
    "swagger": "2.0",
    "securityDefinitions": {
      "oidc": {
        "type": "oauth2",
        "flows": {
          "authorizationCode": {
            "authorizationUrl": "{authorizationUrl}",
            "tokenUrl": "{tokenUrl}",
            "scopes": {
              "openid": "openId scope"
            }
          }
        }
      }
    }
  }
}
```

When I click the `Authorize` button within the API portal, I receive the following message. 

<img width="623" alt="image" src="https://github.com/user-attachments/assets/206ec9f6-107f-4261-9065-d6ffc9274fd5">

Is the authorization code flow supported in the latest version of the API portal for this experience? Or do I need to modify this baseSwagger implementation in some way to meet the expectations of the API portal? 

Let me know what additional information you need. 

Thanks!

For what it is worth, if I attempt to download the open api schema from the api portal it produces the following json. You'll see the scopes and flows are incorrectly formatted: 
```json
{"openapi":"3.0.1","info":{"version":"65c687086e07ab59d311e653","title":"PhilsPDT","description":""},"paths":{"/test/ready":{"get":{"tags":["Health"],"responses":{"200":{"description":"Success"}}}},"/test/-/ready":{"get":{"tags":["Health"],"responses":{"200":{"description":"Success"}}}},"/test/self":{"get":{"tags":["Health"],"responses":{"200":{"description":"Success"}}}},"/test/-/healthz":{"get":{"tags":["Health"],"responses":{"200":{"description":"Success"}}}},"/test/weather-forecast":{"get":{"tags":["WeatherForecast"],"responses":{"200":{"description":"Success","content":{"text/plain":{"schema":{"type":"array","items":{"$ref":"#/components/schemas/App.Controllers.WeatherForecast"}}},"application/json":{"schema":{"type":"array","items":{"$ref":"#/components/schemas/App.Controllers.WeatherForecast"}}},"text/json":{"schema":{"type":"array","items":{"$ref":"#/components/schemas/App.Controllers.WeatherForecast"}}},"application/xml":{"schema":{"type":"array","items":{"$ref":"#/components/schemas/App.Controllers.WeatherForecast"}}},"text/xml":{"schema":{"type":"array","items":{"$ref":"#/components/schemas/App.Controllers.WeatherForecast"}}}}}},"security":[{"oidc":["openid"]}]}},"/test/weather-forecast/without-token":{"get":{"tags":["WeatherForecast"],"responses":{"200":{"description":"Success"}}}}},"components":{"securitySchemes":{"oidc":{"type":"oauth2","flows":{"undefined":{"scopes":{}}}}},"schemas":{"App.Controllers.WeatherForecast":{"type":"object","properties":{"date":{"type":"string","format":"date-time"},"temperatureC":{"type":"integer","format":"int32"},"temperatureF":{"type":"integer","format":"int32","readOnly":true},"summary":{"type":"string","nullable":true}},"additionalProperties":false}}},"servers":[{"url":"/"}],"tags":[]}
```

### Actual Outcome

I am unable to use the `Authorize` button to facilitate the capture of an authorization token from the API portal experience. 

### Expected Outcome

I am expecting to be able to use the `Authorize` button in order to follow the authorization code flow to retrieve a JWT from our IDP.

**Comment by LiliumCandidum** [comment]
Hi @phil-hotchkiss, we replicated your problem and found out it's related to the Swagger Aggregator microservice.
Thank you for your report, we'll fix it as soon as possible and let you know when it's ready.

**Comment by LiliumCandidum** [comment]
Hi again @phil-hotchkiss, I found out what's the problem and how to solve it.

The `baseSwagger` object accepts a swagger v2 configuration but your `securityDefinitions` object is using the OpenAPI v3 schema.
You should follow the swagger v2 schema that you can find here: https://github.com/OAI/OpenAPI-Specification/blob/main/versions/2.0.md#security-scheme-object

So, your `baseSwagger` should become something like this:

```json
{
  "baseSwagger": {
    "swagger": "2.0",
    "securityDefinitions": {
      "oidc": {
          "authorizationUrl": "{authorizationUrl}",
          "flow": "authorizationCode",
          "scopes": {
            "openid": "openId scope",
          },
          "tokenUrl": "{tokenUrl}",
          "type": "oauth2",
      }
    }
  }
}
```

Moreover, I noticed that you used curly brakets in your configuration. If your intention was to interpolate those variables, keep in mind that you should use double curly brakets, like `{{tokenUrl}}`.

Let me know if it's solved or still have problems!

**Reply by phil-hotchkiss** [reply]
This did render the Available Authorizations popup where it recognized the flow. However, I'm still unable to use the authorization code pkce flow. It's asking for a client id and a secret (secret isn't needed in the pkce flow). When I enter the client id and click `Authorize`, I successfully authenticate with my identity provider but it opens the API portal in a new window with no authorization token. Is there maybe an issue with the oauth 2.0 interpretation here? IIRC, the Swagger 2.0 spec only supported up to the implicit authorization flow. Does the base swagger support a swagger 3.0 spec where the authorization code pkce flow is supported? 

As a side note, do we have any ways to pre-populate the client id?

**Reply by LiliumCandidum** [reply]
I don't think it's a 2.0 spec issue because reading their spec it seems supported and by the way, if you configure the API Portal to show the v3.0 or v3.1, it converts the baseSwagger to the correct version.
The baseSwagger supports only the 2.0 spec for now.

We'll investigate further and let you know, even regarding the client id.

N.B. There was an error in my snippet and instead of `"flow": "authorizationCode",`, it should be `"flow": "accessCode",`

**Reply by phil-hotchkiss** [reply]
Changing the flow to `accessCode` still results in a new window opening after successful authorization where the api portal isn't functioning as expected. I have not been successful at getting the authorization token to propagate back to the original page. It opens a new window altogether for the auth and then redirects back to the oauth2-redirect.html page with the following error in the url: `.../oauth2-redirect.html?state=TW9uIEp1bCAxNSAyMDI0IDE1OjM0OjQ4IEdNVC0wNTAwIChDZW50cmFsIERheWxpZ2h0IFRpbWUp&error=invalid_request&error_description=PKCE+code+challenge+is+required+by+the+application`. Is there a way to inform the api portal that the intended flow is the auth code + pkce flow?

**Reply by epessina** [reply]
Hi @phil-hotchkiss , we need further testing to understand the problem and any possible solution. We will come back to you as soon as we have more information

**Comment by LiliumCandidum** [comment, marked as answer]
Hi @phil-hotchkiss,
We should have solved the problem but we would like you to give it a try.
We released the `2.1.0-rc1`API Portal version where we hard-coded an option to enable the pkce flow (but it will be configurable).

Please try it and let us know!

**Reply by phil-hotchkiss** [reply]
Worked like a charm! Thank you!

**Reply by LiliumCandidum** [reply]
Perfect, I'm glad to hear that! Thank you for trying it.
We'll make it configurable and release a stable version asap :blush:

**Comment by epessina** [comment]
Hi @phil-hotchkiss , we have just released a stable version of the services with OAuth 2.0 customization support (as explained [here](https://next.docs.mia-platform.eu/docs/runtime_suite_applications/api-documentation-aggregator/faqs#how-can-i-configure-my-oauth-20-flow)).

Please, let me know if everything works as expected
